File: career/frontend/types/simulation.ts
// Simulation Types - matches simulation-service (port 8013) schemas

export type ScenarioType =
  | 'career_path'        // 진로 경로 시뮬레이션
  | 'skill_development'  // 스킬 개발 시나리오
  | 'course_selection'   // 수강 신청 시나리오
  | 'opportunity'        // 기회 선택 시나리오
  | 'timeline';          // 타임라인 시뮬레이션

export interface ScenarioVariable {
  name: string;
  current_value: unknown;
  simulated_value: unknown;
  impact_description?: string;
}

export interface ScenarioCreate {
  student_id: string;
  scenario_type: ScenarioType;
  name: string;
  description?: string;
  variables: ScenarioVariable[];
  target_date?: string;
}

export interface SimulationResult {
  metric_name: string;
  current_value: number;
  simulated_value: number;
  change_percent: number;
  impact_level: 'positive' | 'negative' | 'neutral';
  explanation?: string;
}

export interface AIAnalysis {
  summary: string;
  strengths: string[];
  risks: string[];
  recommendations: string[];
  next_steps: string[];
  confidence_reason: string;
}

export interface ScenarioResponse {
  scenario_id: string;
  student_id: string;
  scenario_type: ScenarioType;
  name: string;
  description?: string;
  variables: ScenarioVariable[];
  results: SimulationResult[];
  overall_impact_score: number;
  recommendation?: string;
  ai_analysis?: AIAnalysis | null;
  created_at: string;
  is_saved: boolean;
}

export interface ComparisonRequest {
  student_id: string;
  scenario_ids: string[];  // 2-5 scenarios
}

export interface ComparisonResponse {
  comparison_id: string;
  student_id: string;
  scenarios: ScenarioResponse[];
  comparison_matrix: Record<string, Record<string, number>>;
  best_scenario_id: string;
  recommendation: string;
  created_at: string;
}

export interface CareerPathSimulation {
  student_id: string;
  target_role_cd: string;
  current_skills: Record<string, number>;
  planned_courses: string[];
  planned_opportunities: string[];
  timeline_months: number;
}

export interface SkillDevelopmentSimulation {
  student_id: string;
  target_skills: Record<string, number>;  // skill_cd -> target_level
  available_hours_per_week: number;
  learning_resources: string[];
}

// Frontend-specific types
export interface SimulationVariable {
  name: string;
  display_name: string;
  type: 'string' | 'number' | 'select';
  value: string | number;
  options?: string[];
  min?: number;
  max?: number;
}

// Alias for SimulationResult used in UI
export type SimulationResultResponse = ScenarioResponse;

// Scenario update request
export interface ScenarioUpdate {
  name?: string;
  description?: string;
  variables?: ScenarioVariable[];
  target_date?: string;
}

// API Response Types
export interface ScenarioTypeInfo {
  value: string;
  label: string;
  description: string;
}

export interface SuggestedScenario {
  scenario_type: ScenarioType;
  title: string;
  description: string;
  reason: string;
  variables: ScenarioVariable[];
}

// Helper
export const SCENARIO_TYPE_LABELS: Record<ScenarioType, string> = {
  career_path: '진로 경로',
  skill_development: '스킬 개발',
  course_selection: '수강 계획',
  opportunity: '기회 탐색',
  timeline: '타임라인',
};

export const SCENARIO_TYPE_DESCRIPTIONS: Record<ScenarioType, string> = {
  career_path: '목표 직무까지의 경로를 시뮬레이션합니다',
  skill_development: '스킬 개발 시나리오를 비교합니다',
  course_selection: '수강 신청 조합의 효과를 분석합니다',
  opportunity: '다양한 기회 선택의 영향을 예측합니다',
  timeline: '시간에 따른 성장을 시뮬레이션합니다',
};

export const IMPACT_LEVEL_COLORS: Record<string, string> = {
  positive: '#22c55e',  // green-500
  negative: '#ef4444',  // red-500
  neutral: '#6b7280',   // gray-500
};

File: career/services/simulation-service/app/schemas/simulation.py
from pydantic import BaseModel, Field
from typing import Optional, List, Dict, Any
from datetime import date, datetime
from uuid import UUID
from enum import Enum


class ScenarioType(str, Enum):
    CAREER_PATH = "career_path"           # 진로 경로 시뮬레이션
    SKILL_DEVELOPMENT = "skill_development"  # 스킬 개발 시나리오
    COURSE_SELECTION = "course_selection"  # 수강 신청 시나리오
    OPPORTUNITY = "opportunity"           # 기회 선택 시나리오
    TIMELINE = "timeline"                 # 타임라인 시뮬레이션
    ACTIVITY = "activity"                 # 활동 시나리오
    # Legacy values for backward compatibility
    CAREER = "career"                     # Legacy: 진로 시뮬레이션
    SKILL = "skill"                       # Legacy: 스킬 시뮬레이션
    COURSE = "course"                     # Legacy: 수강 시뮬레이션


class ScenarioVariable(BaseModel):
    name: str
    current_value: Any
    simulated_value: Any
    impact_description: Optional[str] = None


class ScenarioCreate(BaseModel):
    student_id: str
    scenario_type: ScenarioType
    name: str
    description: Optional[str] = None
    variables: List[ScenarioVariable]
    target_date: Optional[date] = None


class SimulationResult(BaseModel):
    metric_name: str
    current_value: float
    simulated_value: float
    change_percent: float
    impact_level: str = Field(description="positive, negative, neutral")
    explanation: Optional[str] = None


class ScenarioResponse(BaseModel):
    scenario_id: UUID
    student_id: str
    scenario_type: ScenarioType
    name: str
    description: Optional[str] = None
    variables: List[ScenarioVariable]
    results: List[SimulationResult] = Field(default_factory=list)
    overall_impact_score: float = 0
    recommendation: Optional[str] = None
    ai_analysis: Optional[Dict[str, Any]] = None  # AI 분석 결과
    created_at: datetime
    is_saved: bool = False

    class Config:
        from_attributes = True


class ComparisonRequest(BaseModel):
    student_id: str
    scenario_ids: List[UUID] = Field(min_length=2, max_length=5)


class ComparisonResponse(BaseModel):
    comparison_id: UUID
    student_id: str
    scenarios: List[ScenarioResponse]
    comparison_matrix: Dict[str, Dict[str, float]]
    best_scenario_id: UUID
    recommendation: str
    created_at: datetime


class CareerPathSimulation(BaseModel):
    student_id: str
    target_role_cd: str
    current_skills: Dict[str, int] = Field(default_factory=dict)
    planned_courses: List[str] = Field(default_factory=list)
    planned_opportunities: List[str] = Field(default_factory=list)
    timeline_months: int = 24


class SkillDevelopmentSimulation(BaseModel):
    student_id: str
    target_skills: Dict[str, int]  # skill_cd -> target_level
    available_hours_per_week: int = 10
    learning_resources: List[str] = Field(default_factory=list)


class CourseSelectionSimulation(BaseModel):
    student_id: str
    course_codes: List[str]  # 수강할 과목 코드들
    expected_grades: Dict[str, str] = Field(default_factory=dict)  # course_cd -> expected grade (A, B, C, ...)
    semester: Optional[str] = None  # 학기 (예: "2026-1")


class TimelineSimulation(BaseModel):
    student_id: str
    target_date: date  # 목표 날짜
    milestones: List[Dict[str, Any]] = Field(default_factory=list)  # 중간 목표들
    timeline_type: str = "graduation"  # graduation, job_ready, skill_complete


class ScenarioTypeInfo(BaseModel):
    value: str
    label: str
    description: str


class VariableGuide(BaseModel):
    """시나리오 변수 설정 가이드"""
    name: str
    label: str
    description: str
    input_type: str = "text"  # text, number, select, multi-select
    options: Optional[List[Dict[str, Any]]] = None  # select/multi-select용 옵션
    default_value: Optional[Any] = None
    min_value: Optional[float] = None
    max_value: Optional[float] = None
    required: bool = True


class MetricExplanation(BaseModel):
    """시뮬레이션 결과 메트릭 설명"""
    metric_name: str
    description: str
    unit: str  # %, 점, 시간, 개월 등
    interpretation: str  # 수치 해석 방법
    good_threshold: Optional[float] = None  # 좋은 결과 기준값
    warning_threshold: Optional[float] = None  # 주의 필요 기준값


class ScenarioGuide(BaseModel):
    """시나리오 유형별 전체 가이드"""
    scenario_type: str
    title: str
    description: str
    how_it_works: str  # 시뮬레이션 작동 방식 설명
    steps: List[str]  # 진행 단계
    variables: List[VariableGuide]  # 설정 가능한 변수들
    expected_metrics: List[MetricExplanation]  # 예상되는 결과 메트릭들
    tips: List[str]  # 사용 팁


class AIAnalysis(BaseModel):
    """AI 기반 분석 결과"""
    summary: str  # 핵심 요약
    strengths: List[str]  # 이 시나리오의 장점
    risks: List[str]  # 주의해야 할 위험요소
    recommendations: List[str]  # AI 추천 행동
    next_steps: List[str]  # 다음 단계 제안
    confidence_reason: str  # 종합점수 산정 이유


class SuggestedScenario(BaseModel):
    scenario_type: ScenarioType
    title: str
    description: str
    reason: str  # 추천 이유
    variables: List[ScenarioVariable] = Field(default_factory=list)
